### ISSUE TITLE
Looking for alternative for GetterDict in V2

### ISSUE THREAD

**Issue by vitalik** [pydantic/pydantic#5440]
Hello

[django-ninja](https://github.com/vitalik/django-ninja) uses pydantic to achieve some very nice and highly appreciated functionality like automatically converting ORM-querysets to lists or custom [Resolvers](https://django-ninja.rest-framework.com/guides/response/#resolvers) or [Dot Aliases](https://django-ninja.rest-framework.com/guides/response/#aliases) 

These were achieved by [GetterDict](https://github.com/vitalik/django-ninja/blob/d561e6c3a10c2f83ff1bdba44448afdda02f5831/ninja/schema.py#L41-L86), which is gone in v2

so what I am looking for is some sort of alternative in v2 that can act as a proxy between pydantic-model and ORM-model , to collect values from object attributes (field names and aliases) - like [here where in v1 you have example for xml parsing ](https://docs.pydantic.dev/usage/models/#data-binding)

see also discussion here #5393

Thank you

**Comment by dmontagu** [comment]
@vitalik could you help us understand what you are using GetterDict for that prevents `from_attributes` from being a viable replacement?

Originally, `GetterDict` was just meant as an implementation detail/workaround for making it so pydantic could read attributes with the same API as dict field access. But now with `pydantic_core`, our intent is that that is no longer necessary for the sake of parsing objects into pydantic models via attribute access. If the reason you want `GetterDict` is specifically for pydantic integration, then rather than adding `GetterDict` back into pydantic, I think we'd really prefer to make it so that pydantic_core just automatically does the right thing to get attributes off of an object, or at least handles as many use cases as possible.

I'll also note that as of the next alpha release, which will be released today, you'll be able to use more complex validation aliases for reading data from nested attributes.

Also, if you want `GetterDict` for reasons unrelated to pydantic integration, I think it might make sense to make it into a separate package than pydantic, otherwise we end up needing to maintain it even though it isn't intended for use with the pydantic library.

I'm not saying there isn't a good reason to keep `GetterDict`, or that there isn't a gap left by removing it, but I don't currently understand where the gap lies and I think it would help us help you if you could show us some code/functionality that there isn't a good way to handle in v2 without the `GetterDict`.

**Comment by vitalik** [comment]
Hi @dmontagu @samuelcolvin 

Thank you for looking into this - yeah I'm not saying that we need to keep GetterDict, what I mean is we need some alternative callback function that is called by pydantic when it tries to access any attribute of the object it tries to validate

Please take a look into this screencast that will explain the usage: 
https://vitaliy-public.s3.eu-central-1.amazonaws.com/share/202304/pydantic-ninja.mp4

so basically yeah - if you can hint or provide some hook that pydantic can optionally call when accessing non-dict object - that would solve it

**Comment by vitalik** [comment]
Hi @dmontagu 

Thank you
well yeah - for now GetterDict can be replaced with root_validator(pre)

there are few other issues - but I will open them as separate threads
